src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py

```diff
@@ -303,6 +309,92 @@ def optimize(self, verbose=False):
+            - "counts": the counts of the QAOA circuit output (if the QAOA circuit is evaluated on a QPU or shot-based simulator)
+        """
+
+        # before evaluating the circuit we check that the QAOA object has been compiled
+        if self.compiled == False: raise ValueError("Please compile the QAOA before optimizing it!")
+
+        
+        ## Check the type of the input parameters and save them as a QAOAVariationalBaseParams object at the variable `params_obj`
+
+        # if the parameters are passed as a dictionary we copy and update the variational parameters of the QAOA object
+        if isinstance(params, dict):
+            params_obj = deepcopy(self.variate_params)
+            # we check that the dictionary contains all the parameters of the QAOA object that are not empty
+            for key, value in params_obj.asdict().items():
+                if value.size > 0: 
+                    assert key in params.keys(), f"The parameter `{key}` is missing from the input dictionary"
+            params_obj.update_from_dict(params)
+
+        # if the parameters are passed as a list we copy and update the variational parameters of the QAOA object
+        elif isinstance(params, list) or isinstance(params, np.ndarray):
+            assert len(params) == len(self.variate_params), "The number of parameters does not match the number of parameters in the QAOA circuit"
+            params_obj = deepcopy(self.variate_params)
+            params_obj.update_from_raw(params)
+
+        # if the parameters are passed as a QAOAVariationalBaseParams object we just take it as it is
+        elif isinstance(params, QAOAVariationalBaseParams):
+            params_obj = params
+
+        # if the parameters are passed in a different format, we raise an error
+        else:
+            raise TypeError(f'The input params must be a list or a dictionary. Instead, received {type(params)}')
+        
+        
+        ## Evaluate the QAOA circuit and return the results
+
+        # if the backend is the analytical simulator, we just return the expectation value of the cost Hamiltonian
+        if isinstance(self.backend, QAOABackendAnalyticalSimulator):
+            return {'cost': self.backend.expectation(params_obj)[0]}
+
+        # if the backend is a statevector simulator, we return the expectation value and the uncertainty of the cost Hamiltonian and the statevector
+        if isinstance(self.backend, QAOABaseBackendStatevector):
+            state = self.backend.wavefunction(params_obj)
+            cost, uncertainty = self.backend.expectation_w_uncertainty(params_obj)
+            return {
+                'cost': cost,
+                'uncertainty': uncertainty,
+                'state': state
+            }
+
+        # if the backend is a QPU or a shot-based simulator, we return the expectation value and the uncertainty of the cost Hamiltonian and the counts
```

Comment:
why don't we use `expectation_w_uncertainty` method for non-statevector backends too?